File: src/main/java/org/cubexmc/metro/gui/controller/LineSettingsController.java
package org.cubexmc.metro.gui.controller;

import java.util.Map;
import java.util.UUID;

import org.bukkit.entity.Player;
import org.cubexmc.metro.Metro;
import org.cubexmc.metro.gui.ChatInputManager;
import org.cubexmc.metro.gui.GuiHolder;
import org.cubexmc.metro.gui.GuiSlots;
import org.cubexmc.metro.manager.LanguageManager;
import org.cubexmc.metro.manager.RouteRecorder;
import org.cubexmc.metro.model.Line;
import org.cubexmc.metro.service.LineCommandService;

public final class LineSettingsController {
    private final Metro plugin;
    private final LineCommandService lineService;
    private final GuiPermissionGuard permissionGuard;

    public LineSettingsController(Metro plugin) {
        this.plugin = plugin;
        this.lineService = new LineCommandService(plugin.getLineManager());
        this.permissionGuard = new GuiPermissionGuard(plugin);
    }

    public void handleClick(Player player, GuiHolder holder, int slot) {
        String lineId = holder.getData("lineId");
        Line line = plugin.getLineManager().getLine(lineId);
        if (line == null) {
            player.closeInventory();
            return;
        }

        GuiHolder.GuiView previousView = holder.getPreviousView();
        switch (slot) {
            case GuiSlots.LINE_SETTINGS_ROUTE_RECORDING -> {
                if (!permissionGuard.requireManageLine(player, line)) {
                    player.closeInventory();
                    return;
                }
                handleRouteRecordingToggle(player, line);
                plugin.getGuiManager().openLineSettings(player, lineId, previousView);
            }
            case GuiSlots.LINE_SETTINGS_ROUTE_INFO -> {
                sendRouteInfo(player, line);
                plugin.getGuiManager().openLineSettings(player, lineId, previousView);
            }
            case GuiSlots.LINE_SETTINGS_CLEAR_ROUTE -> {
                if (!permissionGuard.requireManageLine(player, line)) {
                    player.closeInventory();
                    return;
                }
                plugin.getGuiManager().openConfirmAction(player, "CLEAR_ROUTE",
                        lineId, line.getName(), null, 0, holder.snapshot());
            }
            case GuiSlots.LINE_SETTINGS_RAIL_PROTECTION -> {
                if (!permissionGuard.requireManageLine(player, line)) {
                    player.closeInventory();
                    return;
                }
                toggleRailProtection(player, line);
                plugin.getGuiManager().openLineSettings(player, lineId, previousView);
            }
            case GuiSlots.LINE_SETTINGS_RENAME -> requestLineRename(player, line, previousView);
            case GuiSlots.LINE_SETTINGS_MAX_SPEED -> requestLineSpeed(player, line, previousView);
            case GuiSlots.LINE_SETTINGS_TICKET_PRICE -> requestLinePrice(player, line, previousView);
            case GuiSlots.LINE_SETTINGS_CLONE_REVERSE -> requestCloneReverse(player, line, previousView);
            case GuiSlots.LINE_SETTINGS_DELETE -> {
                if (!permissionGuard.requireManageLine(player, line)) {
                    player.closeInventory();
                    return;
                }
                plugin.getGuiManager().openConfirmAction(player, "DELETE_LINE",
                        lineId, line.getName(), null, 0, holder.snapshot());
            }
            case GuiSlots.LINE_SETTINGS_COLOR -> requestLineColor(player, line, previousView);
            case GuiSlots.LINE_SETTINGS_TERMINUS -> requestLineTerminus(player, line, previousView);
            case GuiSlots.LINE_SETTINGS_BACK -> plugin.getGuiManager().openPreviousView(player, holder,
                    () -> plugin.getGuiManager().openLineList(player, 0, false));
            default -> {
            }
        }
    }

    private void toggleRailProtection(Player player, Line line) {
        String lineId = line.getId();
        boolean enabled = !line.isRailProtected();
        if (lineService.setRailProtected(lineId, enabled) == LineCommandService.WriteStatus.SUCCESS) {
            player.sendMessage(plugin.getLanguageManager().getMessage("line.protect_updated",
                    args("line_id", lineId,
                            "state", plugin.getLanguageManager().getMessage(enabled
                                    ? "line.protect_state_enabled"
                                    : "line.protect_state_disabled"))));
        } else {
            player.sendMessage(plugin.getLanguageManager().getMessage("line.protect_update_fail",
                    LanguageManager.put(LanguageManager.args(), "line_id", lineId)));
        }
    }

    private void requestLineRename(Player player, Line line, GuiHolder.GuiView previousView) {
        if (!permissionGuard.requireManageLine(player, line)) {
            player.closeInventory();
            return;
        }
        String lineId = line.getId();
        String oldName = line.getName();
        plugin.getChatInputManager().requestInput(player, plugin.getLanguageManager().getMessage("chat.enter_new_name"),
                new ChatInputManager.ChatInputCallback() {
                    @Override
                    public void onInput(String input) {
                        Line currentLine = requireCurrentLine(player, lineId);
                        if (currentLine == null) {
                            return;
                        }
                        if (lineService.renameLine(lineId, input) == LineCommandService.WriteStatus.SUCCESS) {
                            player.sendMessage(plugin.getLanguageManager().getMessage("line.rename_success",
                                    args("old_name", oldName, "line_id", lineId, "new_name", input)));
                        } else {
                            player.sendMessage(plugin.getLanguageManager().getMessage("line.rename_fail"));
                        }
                        plugin.getGuiManager().openLineSettings(player, lineId, previousView);
                    }

                    @Override
                    public void onCancel() {
                        plugin.getGuiManager().openLineSettings(player, lineId, previousView);
                    }
                });
    }

    private void requestLineSpeed(Player player, Line line, GuiHolder.GuiView previousView) {
        if (!permissionGuard.requireManageLine(player, line)) {
            player.closeInventory();
            return;
        }
        String lineId = line.getId();
        plugin.getChatInputManager().requestInput(player, plugin.getLanguageManager().getMessage("chat.enter_new_speed"),
                new ChatInputManager.ChatInputCallback() {
                    @Override
                    public void onInput(String input) {
                        if (requireCurrentLine(player, lineId) == null) {
                            return;
                        }
                        try {
                            double speed = Double.parseDouble(input);
                            LineCommandService.WriteStatus status = lineService.setMaxSpeed(lineId, speed);
                            if (status == LineCommandService.WriteStatus.INVALID_VALUE) {
                                throw new NumberFormatException();
                            }
                            if (status == LineCommandService.WriteStatus.SUCCESS) {
                                player.sendMessage(plugin.getLanguageManager().getMessage("line.setmaxspeed_success",
                                        args("line_id", lineId, "max_speed", String.valueOf(speed))));
                            }
                        } catch (NumberFormatException e) {
                            player.sendMessage(plugin.getLanguageManager().getMessage("line.setmaxspeed_invalid"));
                        }
                        plugin.getGuiManager().openLineSettings(player, lineId, previousView);
                    }

                    @Override
                    public void onCancel() {
                        plugin.getGuiManager().openLineSettings(player, lineId, previousView);
                    }
                });
    }

    private void requestLinePrice(Player player, Line line, GuiHolder.GuiView previousView) {
        if (!permissionGuard.requireManageLine(player, line)) {
            player.closeInventory();
            return;
        }
        String lineId = line.getId();
        plugin.getChatInputManager().requestInput(player, plugin.getLanguageManager().getMessage("chat.enter_new_price"),
                new ChatInputManager.ChatInputCallback() {
                    @Override
                    public void onInput(String input) {
                        Line currentLine = requireCurrentLine(player, lineId);
                        if (currentLine == null) {
                            return;
                        }
                        try {
                            double price = Double.parseDouble(input);
                            LineCommandService.WriteStatus status = lineService.setTicketPrice(lineId, price);
                            if (status == LineCommandService.WriteStatus.INVALID_VALUE) {
                                throw new NumberFormatException();
                            }
                            if (status == LineCommandService.WriteStatus.SUCCESS) {
                                player.sendMessage(plugin.getLanguageManager().getMessage("line.setprice_success",
                                        args("line_name", currentLine.getName(), "price", String.valueOf(price))));
                            } else {
                                player.sendMessage(plugin.getLanguageManager().getMessage("line.setprice_fail"));
                            }
                        } catch (NumberFormatException e) {
                            player.sendMessage(plugin.getLanguageManager().getMessage("line.setprice_invalid"));
                        }
                        plugin.getGuiManager().openLineSettings(player, lineId, previousView);
                    }

                    @Override
                    public void onCancel() {
                        plugin.getGuiManager().openLineSettings(player, lineId, previousView);
                    }
                });
    }

    private void requestCloneReverse(Player player, Line line, GuiHolder.GuiView previousView) {
        if (!permissionGuard.requireManageLine(player, line) || !permissionGuard.requireCreateLine(player)) {
            player.closeInventory();
            return;
        }
        String lineId = line.getId();
        plugin.getChatInputManager().requestInput(player, plugin.getLanguageManager().getMessage("chat.enter_clone_info"),
                new ChatInputManager.ChatInputCallback() {
                    @Override
                    public void onInput(String input) {
                        if (requireCurrentLine(player, lineId) == null || !permissionGuard.requireCreateLine(player)) {
                            player.closeInventory();
                            return;
                        }
                        String[] parts = input.split(" ");
                        if (parts.length >= 1 && !parts[0].isEmpty()) {
                            String newId = parts[0];
                            String stopIdSuffix = parts.length > 1 ? parts[1] : "_rev";
                            LineCommandService.WriteStatus status = lineService.cloneReverseLine(lineId, newId,
                                    stopIdSuffix, player.getUniqueId());
                            if (status == LineCommandService.WriteStatus.SUCCESS) {
                                player.sendMessage(plugin.getLanguageManager().getMessage("line.clone_success",
                                        LanguageManager.put(LanguageManager.args(), "new_line_id", newId)));
                            } else {
                                player.sendMessage(plugin.getLanguageManager().getMessage("line.clone_fail"));
                            }
                        }
                        plugin.getGuiManager().openLineSettings(player, lineId, previousView);
                    }

                    @Override
                    public void onCancel() {
                        plugin.getGuiManager().openLineSettings(player, lineId, previousView);
                    }
                });
    }

    private void requestLineColor(Player player, Line line, GuiHolder.GuiView previousView) {
        if (!permissionGuard.requireManageLine(player, line)) {
            player.closeInventory();
            return;
        }
        String lineId = line.getId();
        plugin.getChatInputManager().requestInput(player, plugin.getLanguageManager().getMessage("chat.enter_new_color"),
                new ChatInputManager.ChatInputCallback() {
                    @Override
                    public void onInput(String input) {
                        Line currentLine = requireCurrentLine(player, lineId);
                        if (currentLine == null) {
                            return;
                        }
                        String color = input.trim();
                        LineCommandService.WriteStatus status = lineService.setColor(lineId, color);
                        if (status == LineCommandService.WriteStatus.INVALID_COLOR) {
                            player.sendMessage(plugin.getLanguageManager().getMessage("line.setcolor_invalid",
                                    LanguageManager.put(LanguageManager.args(), "color", color)));
                        } else if (status == LineCommandService.WriteStatus.SUCCESS) {
                            player.sendMessage(plugin.getLanguageManager().getMessage("line.setcolor_success",
                                    args("line_id", lineId, "line_name", currentLine.getName(), "color", color)));
                        } else {
                            player.sendMessage(plugin.getLanguageManager().getMessage("line.setcolor_fail"));
                        }
                        plugin.getGuiManager().openLineSettings(player, lineId, previousView);
                    }

                    @Override
                    public void onCancel() {
                        plugin.getGuiManager().openLineSettings(player, lineId, previousView);
                    }
                });
    }

    private void requestLineTerminus(Player player, Line line, GuiHolder.GuiView previousView) {
        if (!permissionGuard.requireManageLine(player, line)) {
            player.closeInventory();
            return;
        }
        String lineId = line.getId();
        plugin.getChatInputManager().requestInput(player,
                plugin.getLanguageManager().getMessage("chat.enter_new_terminus"),
                new ChatInputManager.ChatInputCallback() {
                    @Override
                    public void onInput(String input) {
                        if (requireCurrentLine(player, lineId) == null) {
                            return;
                        }
                        String terminusName = input.trim();
                        if (lineService.setTerminusName(lineId, terminusName)
                                == LineCommandService.WriteStatus.SUCCESS) {
                            player.sendMessage(plugin.getLanguageManager().getMessage("line.setterminus_success",
                                    args("line_id", lineId, "terminus_name", terminusName)));
                        } else {
                            player.sendMessage(plugin.getLanguageManager().getMessage("line.setterminus_fail"));
                        }
                        plugin.getGuiManager().openLineSettings(player, lineId, previousView);
                    }

                    @Override
                    public void onCancel() {
                        plugin.getGuiManager().openLineSettings(player, lineId, previousView);
                    }
                });
    }

    private Line requireCurrentLine(Player player, String lineId) {
        Line currentLine = plugin.getLineManager().getLine(lineId);
        if (currentLine == null) {
            player.closeInventory();
            return null;
        }
        if (!permissionGuard.requireManageLine(player, currentLine)) {
            player.closeInventory();
            return null;
        }
        return currentLine;
    }

    private void handleRouteRecordingToggle(Player player, Line line) {
        String lineId = line.getId();
        RouteRecorder recorder = plugin.getRouteRecorder();
        if (recorder.isRecording(lineId)) {
            RouteRecorder.FinishResult result = recorder.stopAndSave(lineId);
            switch (result.status()) {
                case SAVED -> player.sendMessage(plugin.getLanguageManager().getMessage("line.record_saved",
                        args("line_id", lineId, "point_count", String.valueOf(result.pointCount()))));
                case TOO_FEW_POINTS -> sendRecordTooFew(player, result);
                case FAILED -> player.sendMessage(plugin.getLanguageManager().getMessage("line.record_failed",
                        LanguageManager.put(LanguageManager.args(), "line_id", lineId)));
                case NOT_RECORDING -> player.sendMessage(plugin.getLanguageManager().getMessage("line.record_not_recording",
                        LanguageManager.put(LanguageManager.args(), "line_id", lineId)));
            }
            return;
        }

        if (recorder.start(lineId)) {
            player.sendMessage(plugin.getLanguageManager().getMessage("line.record_started",
                    LanguageManager.put(LanguageManager.args(), "line_id", lineId)));
            player.sendMessage(plugin.getLanguageManager().getMessage("line.record_hint"));
        } else {
            player.sendMessage(plugin.getLanguageManager().getMessage("line.record_already",
                    LanguageManager.put(LanguageManager.args(), "line_id", lineId)));
        }
    }

    private void sendRouteInfo(Player player, Line line) {
        RouteRecorder recorder = plugin.getRouteRecorder();
        player.sendMessage(plugin.getLanguageManager().getMessage("line.routeinfo_header",
                args("line_name", line.getName(), "line_id", line.getId())));
        player.sendMessage(plugin.getLanguageManager().getMessage("line.routeinfo_saved_points",
                LanguageManager.put(LanguageManager.args(), "point_count", String.valueOf(line.getRoutePoints().size()))));
        player.sendMessage(plugin.getLanguageManager().getMessage("line.protect_status",
                LanguageManager.put(LanguageManager.args(), "state", plugin.getLanguageManager().getMessage(line.isRailProtected()
                        ? "line.protect_state_enabled"
                        : "line.protect_state_disabled"))));
        int protectedBlocks = plugin.getRailProtectionManager() == null
                ? 0
                : plugin.getRailProtectionManager().getProtectedBlockCount(line.getId());
        player.sendMessage(plugin.getLanguageManager().getMessage("line.protect_blocks",
                LanguageManager.put(LanguageManager.args(), "count", String.valueOf(protectedBlocks))));
        if (line.isRailProtected() && protectedBlocks == 0) {
            player.sendMessage(plugin.getLanguageManager().getMessage("line.protect_no_blocks"));
        }
        if (recorder.isRecording(line.getId())) {
            UUID cartId = recorder.getRecordingCartId(line.getId());
            player.sendMessage(plugin.getLanguageManager().getMessage("line.routeinfo_recording",
                    LanguageManager.put(LanguageManager.args(), "state",
                            plugin.getLanguageManager().getMessage("line.routeinfo_recording_active"))));
            player.sendMessage(plugin.getLanguageManager().getMessage("line.routeinfo_buffered_points",
                    LanguageManager.put(LanguageManager.args(), "point_count",
                            String.valueOf(recorder.getActivePointCount(line.getId())))));
            player.sendMessage(plugin.getLanguageManager().getMessage("line.routeinfo_bound_cart",
                    LanguageManager.put(LanguageManager.args(), "cart_id",
                            cartId == null
                                    ? plugin.getLanguageManager().getMessage("line.routeinfo_waiting_cart")
                                    : cartId.toString())));
        } else {
            player.sendMessage(plugin.getLanguageManager().getMessage("line.routeinfo_recording",
                    LanguageManager.put(LanguageManager.args(), "state",
                            plugin.getLanguageManager().getMessage("line.routeinfo_recording_inactive"))));
        }
    }

    private Map<String, Object> args(Object... replacements) {
        Map<String, Object> args = LanguageManager.args();
        for (int i = 0; i < replacements.length - 1; i += 2) {
            LanguageManager.put(args, String.valueOf(replacements[i]), replacements[i + 1]);
        }
        return args;
    }

    private void sendRecordTooFew(Player player, RouteRecorder.FinishResult result) {
        player.sendMessage(plugin.getLanguageManager().getMessage("line.record_too_few",
                LanguageManager.put(LanguageManager.args(), "point_count", String.valueOf(result.pointCount()))));
        player.sendMessage(plugin.getLanguageManager().getMessage("line.record_too_few_hint"));
    }
}

File: src/main/java/org/cubexmc/metro/gui/controller/LineSettingsController.kt
package org.cubexmc.metro.gui.controller

import org.bukkit.entity.Player
import org.cubexmc.metro.Metro
import org.cubexmc.metro.gui.ChatInputManager
import org.cubexmc.metro.gui.GuiHolder
import org.cubexmc.metro.gui.GuiSlots
import org.cubexmc.metro.manager.LanguageManager
import org.cubexmc.metro.manager.RouteRecorder
import org.cubexmc.metro.model.Line
import org.cubexmc.metro.service.LineCommandService

class LineSettingsController(private val plugin: Metro) {

    private val lineService = LineCommandService(plugin.lineManager)
    private val permissionGuard = GuiPermissionGuard(plugin)

    fun handleClick(player: Player, holder: GuiHolder, slot: Int) {
        val lineId = holder.getData<String>("lineId")
        val line = if (lineId == null) null else plugin.lineManager.getLine(lineId)
        if (line == null) {
            player.closeInventory()
            return
        }

        val previousView = holder.getPreviousView()
        when (slot) {
            GuiSlots.LINE_SETTINGS_ROUTE_RECORDING -> {
                if (!permissionGuard.requireManageLine(player, line)) {
                    player.closeInventory()
                    return
                }
                handleRouteRecordingToggle(player, line)
                plugin.guiManager.openLineSettings(player, lineId, previousView)
            }

            GuiSlots.LINE_SETTINGS_ROUTE_INFO -> {
                sendRouteInfo(player, line)
                plugin.guiManager.openLineSettings(player, lineId, previousView)
            }

            GuiSlots.LINE_SETTINGS_CLEAR_ROUTE -> {
                if (!permissionGuard.requireManageLine(player, line)) {
                    player.closeInventory()
                    return
                }
                plugin.guiManager.openConfirmAction(
                    player,
                    "CLEAR_ROUTE",
                    lineId,
                    line.name,
                    null,
                    0,
                    holder.snapshot(),
                )
            }

            GuiSlots.LINE_SETTINGS_RAIL_PROTECTION -> {
                if (!permissionGuard.requireManageLine(player, line)) {
                    player.closeInventory()
                    return
                }
                toggleRailProtection(player, line)
                plugin.guiManager.openLineSettings(player, lineId, previousView)
            }

            GuiSlots.LINE_SETTINGS_RENAME -> requestLineRename(player, line, previousView)

            GuiSlots.LINE_SETTINGS_MAX_SPEED -> requestLineSpeed(player, line, previousView)

            GuiSlots.LINE_SETTINGS_TICKET_PRICE -> requestLinePrice(player, line, previousView)

            GuiSlots.LINE_SETTINGS_CLONE_REVERSE -> requestCloneReverse(player, line, previousView)

            GuiSlots.LINE_SETTINGS_DELETE -> {
                if (!permissionGuard.requireManageLine(player, line)) {
                    player.closeInventory()
                    return
                }
                plugin.guiManager.openConfirmAction(
                    player,
                    "DELETE_LINE",
                    lineId,
                    line.name,
                    null,
                    0,
                    holder.snapshot(),
                )
            }

            GuiSlots.LINE_SETTINGS_COLOR -> requestLineColor(player, line, previousView)

            GuiSlots.LINE_SETTINGS_TERMINUS -> requestLineTerminus(player, line, previousView)

            GuiSlots.LINE_SETTINGS_BACK ->
                plugin.guiManager.openPreviousView(player, holder) {
                    plugin.guiManager.openLineList(player, 0, false)
                }

            else -> Unit
        }
    }

    private fun toggleRailProtection(player: Player, line: Line) {
        val lineId = line.id
        val enabled = !line.isRailProtected
        if (lineService.setRailProtected(lineId, enabled) == LineCommandService.WriteStatus.SUCCESS) {
            val stateKey = if (enabled) "line.protect_state_enabled" else "line.protect_state_disabled"
            player.sendMessage(
                plugin.languageManager.getMessage(
                    "line.protect_updated",
                    args("line_id", lineId, "state", plugin.languageManager.getMessage(stateKey)),
                ),
            )
        } else {
            player.sendMessage(
                plugin.languageManager.getMessage(
                    "line.protect_update_fail",
                    LanguageManager.put(LanguageManager.args(), "line_id", lineId),
                ),
            )
        }
    }

    private fun requestLineRename(player: Player, line: Line, previousView: GuiHolder.GuiView?) {
        if (!permissionGuard.requireManageLine(player, line)) {
            player.closeInventory()
            return
        }
        val lineId = line.id
        val oldName = line.name
        requestInput(
            player,
            "chat.enter_new_name",
            lineId,
            previousView,
        ) { input ->
            if (requireCurrentLine(player, lineId) == null) {
                return@requestInput
            }
            if (lineService.renameLine(lineId, input) == LineCommandService.WriteStatus.SUCCESS) {
                player.sendMessage(
                    plugin.languageManager.getMessage(
                        "line.rename_success",
                        args("old_name", oldName, "line_id", lineId, "new_name", input),
                    ),
                )
            } else {
                player.sendMessage(plugin.languageManager.getMessage("line.rename_fail"))
            }
            plugin.guiManager.openLineSettings(player, lineId, previousView)
        }
    }

    private fun requestLineSpeed(player: Player, line: Line, previousView: GuiHolder.GuiView?) {
        if (!permissionGuard.requireManageLine(player, line)) {
            player.closeInventory()
            return
        }
        val lineId = line.id
        requestInput(player, "chat.enter_new_speed", lineId, previousView) { input ->
            if (requireCurrentLine(player, lineId) == null) {
                return@requestInput
            }
            try {
                val speed = input.toDouble()
                val status = lineService.setMaxSpeed(lineId, speed)
                if (status == LineCommandService.WriteStatus.INVALID_VALUE) {
                    throw NumberFormatException()
                }
                if (status == LineCommandService.WriteStatus.SUCCESS) {
                    player.sendMessage(
                        plugin.languageManager.getMessage(
                            "line.setmaxspeed_success",
                            args("line_id", lineId, "max_speed", speed.toString()),
                        ),
                    )
                }
            } catch (_: NumberFormatException) {
                player.sendMessage(plugin.languageManager.getMessage("line.setmaxspeed_invalid"))
            }
            plugin.guiManager.openLineSettings(player, lineId, previousView)
        }
    }

    private fun requestLinePrice(player: Player, line: Line, previousView: GuiHolder.GuiView?) {
        if (!permissionGuard.requireManageLine(player, line)) {
            player.closeInventory()
            return
        }
        val lineId = line.id
        requestInput(player, "chat.enter_new_price", lineId, previousView) { input ->
            val currentLine = requireCurrentLine(player, lineId) ?: return@requestInput
            try {
                val price = input.toDouble()
                val status = lineService.setTicketPrice(lineId, price)
                if (status == LineCommandService.WriteStatus.INVALID_VALUE) {
                    throw NumberFormatException()
                }
                if (status == LineCommandService.WriteStatus.SUCCESS) {
                    player.sendMessage(
                        plugin.languageManager.getMessage(
                            "line.setprice_success",
                            args("line_name", currentLine.name, "price", price.toString()),
                        ),
                    )
                } else {
                    player.sendMessage(plugin.languageManager.getMessage("line.setprice_fail"))
                }
            } catch (_: NumberFormatException) {
                player.sendMessage(plugin.languageManager.getMessage("line.setprice_invalid"))
            }
            plugin.guiManager.openLineSettings(player, lineId, previousView)
        }
    }

    private fun requestCloneReverse(player: Player, line: Line, previousView: GuiHolder.GuiView?) {
        if (!permissionGuard.requireManageLine(player, line) || !permissionGuard.requireCreateLine(player)) {
            player.closeInventory()
            return
        }
        val lineId = line.id
        requestInput(player, "chat.enter_clone_info", lineId, previousView) { input ->
            if (requireCurrentLine(player, lineId) == null || !permissionGuard.requireCreateLine(player)) {
                player.closeInventory()
                return@requestInput
            }
            val parts = input.split(" ")
            if (parts.isNotEmpty() && parts[0].isNotEmpty()) {
                val newId = parts[0]
                val stopIdSuffix = if (parts.size > 1) parts[1] else "_rev"
                val status = lineService.cloneReverseLine(lineId, newId, stopIdSuffix, player.uniqueId)
                if (status == LineCommandService.WriteStatus.SUCCESS) {
                    player.sendMessage(
                        plugin.languageManager.getMessage(
                            "line.clone_success",
                            LanguageManager.put(LanguageManager.args(), "new_line_id", newId),
                        ),
                    )
                } else {
                    player.sendMessage(plugin.languageManager.getMessage("line.clone_fail"))
                }
            }
            plugin.guiManager.openLineSettings(player, lineId, previousView)
        }
    }

    private fun requestLineColor(player: Player, line: Line, previousView: GuiHolder.GuiView?) {
        if (!permissionGuard.requireManageLine(player, line)) {
            player.closeInventory()
            return
        }
        val lineId = line.id
        requestInput(player, "chat.enter_new_color", lineId, previousView) { input ->
            val currentLine = requireCurrentLine(player, lineId) ?: return@requestInput
            val color = input.trim()
            when (lineService.setColor(lineId, color)) {
                LineCommandService.WriteStatus.INVALID_COLOR ->
                    player.sendMessage(
                        plugin.languageManager.getMessage(
                            "line.setcolor_invalid",
                            LanguageManager.put(LanguageManager.args(), "color", color),
                        ),
                    )

                LineCommandService.WriteStatus.SUCCESS ->
                    player.sendMessage(
                        plugin.languageManager.getMessage(
                            "line.setcolor_success",
                            args("line_id", lineId, "line_name", currentLine.name, "color_code", color),
                        ),
                    )

                else -> player.sendMessage(plugin.languageManager.getMessage("line.setcolor_fail"))
            }
            plugin.guiManager.openLineSettings(player, lineId, previousView)
        }
    }

    private fun requestLineTerminus(player: Player, line: Line, previousView: GuiHolder.GuiView?) {
        if (!permissionGuard.requireManageLine(player, line)) {
            player.closeInventory()
            return
        }
        val lineId = line.id
        requestInput(player, "chat.enter_new_terminus", lineId, previousView) { input ->
            if (requireCurrentLine(player, lineId) == null) {
                return@requestInput
            }
            val terminusName = input.trim()
            if (lineService.setTerminusName(lineId, terminusName) == LineCommandService.WriteStatus.SUCCESS) {
                player.sendMessage(
                    plugin.languageManager.getMessage(
                        "line.setterminus_success",
                        args("line_id", lineId, "terminus_name", terminusName),
                    ),
                )
            } else {
                player.sendMessage(plugin.languageManager.getMessage("line.setterminus_fail"))
            }
            plugin.guiManager.openLineSettings(player, lineId, previousView)
        }
    }

    /**
     * Prompts for chat input; cancelling always returns to the line settings view, while each
     * caller decides when to reopen it, mirroring the per-action anonymous callbacks the Java
     * implementation used.
     */
    private fun requestInput(
        player: Player,
        promptKey: String,
        lineId: String,
        previousView: GuiHolder.GuiView?,
        onInput: (String) -> Unit,
    ) {
        plugin.chatInputManager.requestInput(
            player,
            plugin.languageManager.getMessage(promptKey),
            object : ChatInputManager.ChatInputCallback {
                override fun onInput(input: String) {
                    onInput.invoke(input)
                }

                override fun onCancel() {
                    plugin.guiManager.openLineSettings(player, lineId, previousView)
                }
            },
        )
    }

    private fun requireCurrentLine(player: Player, lineId: String): Line? {
        val currentLine = plugin.lineManager.getLine(lineId)
        if (currentLine == null) {
            player.closeInventory()
            return null
        }
        if (!permissionGuard.requireManageLine(player, currentLine)) {
            player.closeInventory()
            return null
        }
        return currentLine
    }

    private fun handleRouteRecordingToggle(player: Player, line: Line) {
        val lineId = line.id
        val recorder = plugin.routeRecorder
        if (recorder.isRecording(lineId)) {
            sendFinishResult(player, lineId, recorder.stopAndSave(lineId))
            return
        }

        if (recorder.start(lineId)) {
            player.sendMessage(
                plugin.languageManager.getMessage(
                    "line.record_started",
                    LanguageManager.put(LanguageManager.args(), "line_id", lineId),
                ),
            )
            player.sendMessage(plugin.languageManager.getMessage("line.record_hint"))
        } else {
            player.sendMessage(
                plugin.languageManager.getMessage(
                    "line.record_already",
                    LanguageManager.put(LanguageManager.args(), "line_id", lineId),
                ),
            )
        }
    }

    private fun sendFinishResult(player: Player, lineId: String, result: RouteRecorder.FinishResult) {
        when (result.status) {
            RouteRecorder.FinishResult.Status.SAVED ->
                player.sendMessage(
                    plugin.languageManager.getMessage(
                        "line.record_saved",
                        args("line_id", lineId, "point_count", result.pointCount.toString()),
                    ),
                )

            RouteRecorder.FinishResult.Status.TOO_FEW_POINTS -> sendRecordTooFew(player, result)

            RouteRecorder.FinishResult.Status.FAILED ->
                player.sendMessage(
                    plugin.languageManager.getMessage(
                        "line.record_failed",
                        LanguageManager.put(LanguageManager.args(), "line_id", lineId),
                    ),
                )

            RouteRecorder.FinishResult.Status.NOT_RECORDING ->
                player.sendMessage(
                    plugin.languageManager.getMessage(
                        "line.record_not_recording",
                        LanguageManager.put(LanguageManager.args(), "line_id", lineId),
                    ),
                )
        }
    }

    private fun sendRouteInfo(player: Player, line: Line) {
        val lang = plugin.languageManager
        val recorder = plugin.routeRecorder
        player.sendMessage(
            lang.getMessage("line.routeinfo_header", args("line_name", line.name, "line_id", line.id)),
        )
        player.sendMessage(
            lang.getMessage(
                "line.routeinfo_saved_points",
                LanguageManager.put(LanguageManager.args(), "point_count", line.routePoints.size.toString()),
            ),
        )
        val protectionStateKey =
            if (line.isRailProtected) "line.protect_state_enabled" else "line.protect_state_disabled"
        player.sendMessage(
            lang.getMessage(
                "line.protect_status",
                LanguageManager.put(LanguageManager.args(), "state", lang.getMessage(protectionStateKey)),
            ),
        )
        val protectedBlocks = plugin.railProtectionManager?.getProtectedBlockCount(line.id) ?: 0
        player.sendMessage(
            lang.getMessage(
                "line.protect_blocks",
                LanguageManager.put(LanguageManager.args(), "count", protectedBlocks.toString()),
            ),
        )
        if (line.isRailProtected && protectedBlocks == 0) {
            player.sendMessage(lang.getMessage("line.protect_no_blocks"))
        }
        sendRecordingState(player, line, recorder)
    }

    private fun sendRecordingState(player: Player, line: Line, recorder: RouteRecorder) {
        val lang = plugin.languageManager
        if (!recorder.isRecording(line.id)) {
            player.sendMessage(
                lang.getMessage(
                    "line.routeinfo_recording",
                    LanguageManager.put(
                        LanguageManager.args(),
                        "state",
                        lang.getMessage("line.routeinfo_recording_inactive"),
                    ),
                ),
            )
            return
        }

        val cartId = recorder.getRecordingCartId(line.id)
        player.sendMessage(
            lang.getMessage(
                "line.routeinfo_recording",
                LanguageManager.put(
                    LanguageManager.args(),
                    "state",
                    lang.getMessage("line.routeinfo_recording_active"),
                ),
            ),
        )
        player.sendMessage(
            lang.getMessage(
                "line.routeinfo_buffered_points",
                LanguageManager.put(
                    LanguageManager.args(),
                    "point_count",
                    recorder.getActivePointCount(line.id).toString(),
                ),
            ),
        )
        player.sendMessage(
            lang.getMessage(
                "line.routeinfo_bound_cart",
                LanguageManager.put(
                    LanguageManager.args(),
                    "cart_id",
                    cartId?.toString() ?: lang.getMessage("line.routeinfo_waiting_cart"),
                ),
            ),
        )
    }

    private fun args(vararg replacements: Any?): MutableMap<String, Any?> {
        val args = LanguageManager.args()
        var index = 0
        while (index < replacements.size - 1) {
            LanguageManager.put(args, replacements[index].toString(), replacements[index + 1])
            index += 2
        }
        return args
    }

    private fun sendRecordTooFew(player: Player, result: RouteRecorder.FinishResult) {
        player.sendMessage(
            plugin.languageManager.getMessage(
                "line.record_too_few",
                LanguageManager.put(LanguageManager.args(), "point_count", result.pointCount.toString()),
            ),
        )
        player.sendMessage(plugin.languageManager.getMessage("line.record_too_few_hint"))
    }
}
